### PR TITLE
Refactor - Merging MUD & MUD Examples - Pt 1

### PR DESCRIPTION
**Pull request by cdelcastillo21**
General Changes
---------------

  * Changes for PEP 8 Standards - line length, lines between functions in class
  * Type hints for function inputs
  * Numpy style documentation 

base.py
-------
* Changes:
  * DensityProblem:
    * Numpy style docs added.
    * Exposed n_params, n_features, n_samples as public properties of class
    * weights behvaior changes:
      * set_weights run on initialization. _weights attribute always set, with default being a vector of ones. 
      * storing of initial, observed, and predicted distribution functions in class, to be re-used in plotting functions.
      * Sacked weights for DensityProblem class. This allows a set of weights to be specified (for incorporating weights from multiple previous iterations) instead of a singular vector of weights.
      * Option to normalize weights vector, set to False by default in set_weights
    * Plotting functions:
      * plot_param_space - For plotting initial, updated, and weighted initial for a single parameter.
      * plot_obs_space - For plotting observed, predicted, and push-forward of updated for a single observable.
      * set_predicted weights behavior - Weights argument, if passed, goes through set_weights first. self._weights always passed to gkde, with the default value of all ones not affecting the result. 
      * exp_r - Added function for computing the expectation of the r ratio, which should be close to one. Implemented as weigthed average of self._r using initial weights.
  * BayesProblem:
    * Minor changes to initialization routine. 
    * Exposed n_params, n_features, n_samples as public properties of class
    * Added plotting routines similar to DensityProblem
* Additions
  * LinearGaussianProblem Class - For solving linear parameter estiamtion problems. Implements MUD, MAP, and Least-Squares solutions, functions to compute regularization functional terms, and methos to plot contours and solutions. More documentation and testing pending.
  * IterativeLinearProblem - Extension of LinearGaussianProblem. Implements iterative algorithm for linear maps, where iterations are taken over different rows of A. Plotting functions included as well, for plotting chain of solutions and different contours of the linear map. 
	
examples.py
-----------
Added file including functions that impelement problem constructors for the different classes in base.py. Functions here used in tests and jupyter notebooks. Current functions included:

  * rotation_map, rotation_map_trials - For IterativeLinearProblem class examples involving rotation_map. Examples demonstrating skewness property of QoI map.

  * identity_uniform_1D_density_prob - Similar to the function in conftest, just added a few more functionalities to it. Used now in test-suite

util.py
-------
* Additions:
  * make_2d_unit_mesh - originally in mud_examples repo, needed for plotting routines in all classes.

contest.py
----------
Changed problem generator to use function from examples.py. Same function and same exmaples are being constructed for all tests. 

test_base.py
------------
  * Minor formatting changes
  * Added tests for normalizing weights, and for plotting routines in DensityProblem class.

test_setting_predicted.py
-------------------------
  * Changes now that properties are public for DensityProblem class.

### CODE REVIEW

**Comment by cdelcastillo21** [comment]
There seems to be a problem importing the numpy.typing class [ArrayLike](https://numpy.org/devdocs/reference/typing.html). Its offered as of version 1.20 in numpy. When I ran the tests locally on my machine this wasn't an issue, but I see in setup.cfg that no specific numpy version is specified, so maybe the unit tests are running under an older version of numpy? 

We can also just discard the use of ArrayLike all together for type hinting and going back to just using Unions of list and numpy.ndarray. I just recently found out about the ArrayLike class and was curious about using it.

**Comment by mathematicalmichael** [comment]
> There seems to be a problem importing the numpy.typing class [ArrayLike](https://numpy.org/devdocs/reference/typing.html). Its offered as of version 1.20 in numpy. When I ran the tests locally on my machine this wasn't an issue, but I see in setup.cfg that no specific numpy version is specified, so maybe the unit tests are running under an older version of numpy? 
> 
> 
> 
> We can also just discard the use of ArrayLike all together for type hinting and going back to just using Unions of list and numpy.ndarray. I just recently found out about the ArrayLike class and was curious about using it. 

Keep the new type hint, I prefer that. Just edit setup.cfg to make this numpy version the minimum required. `numpy~=1.20` or whatever.

**Comment by mathematicalmichael** [comment]
you may want to run `black` or `autopep8` on the code to silence the Hound warnings.

File: src/mud/examples.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+from scipy.stats import distributions as ds
+
+
+def rotation_map(qnum=10, tol=0.1, b=None, ref_param=None, seed=None):
+    """
+    Generate test data linear rotation map
+
+    """
+    if seed is not None:
+        np.random.seed(24)
```
Comment (mathematicalmichael) [inline src/mud/examples.py:15]:
```suggestion
        np.random.seed(seed)
```

File: src/mud/base.py [hunk 2]
```diff
@@ -113,31 +333,312 @@ def fit(self, **kwargs):
...
+        ----------
+
+        Returns
+        -------
+        mud_point : ndarray
+            Maximal Updated Density (MUD) point.
+        """
         return self.mud_point()
 
+    def exp_r(self):
```
Comment (mathematicalmichael) [inline src/mud/base.py:383]: thoughts on calling it `ratio_expected`? would show up next to `ratio` in the doc autocomplete.

Comment (mathematicalmichael) [reply src/mud/base.py:383]: `expected_ratio` is better even.

**Pull request merged**